pkg/util/push/request.go

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+type supplierFunc func() (*mimirpb.WriteRequest, func(), error)
+
+// Request represents a push request. It allows lazy body reading from the underlying http request
+// and adding cleanups that should be done after the request is completed.
+type Request struct {
+	cleanups []func()
+
+	getRequest supplierFunc
+
+	requestIsParsed bool
+	parsedRequest   *mimirpb.WriteRequest
+	parseErr        error
+}
+
+func newRequest(p supplierFunc) *Request {
+	return &Request{
+		cleanups:   make([]func(), 0, 10),
+		getRequest: p,
+	}
+}
+
+func NewParsedRequest(r *mimirpb.WriteRequest) *Request {
+	return newRequest(func() (*mimirpb.WriteRequest, func(), error) {
+		return r, nil, nil
+	})
+}
+
+func (r *Request) parseRequest() {
+	var cleanup func()
+	r.parsedRequest, cleanup, r.parseErr = r.getRequest()
+	r.requestIsParsed = true
+	if cleanup != nil {
+		r.AddCleanup(cleanup)
+	}
+}
+
+// WriteRequest parses the body of the push request and caches it.
+// Subsequent calls to WriteRequest return the same results.
+func (r *Request) WriteRequest() (*mimirpb.WriteRequest, error) {
+	if !r.requestIsParsed {
```

Comment:
it kind of changes the contract with the supplier function. Now it will be invoked only once if it returns a non-nil error or a body. Otherwise it will be invoked on every `WriteRequest()` call. I would prefer to keep it as-is

Reply:
> Otherwise

This means it returned `nil` request and error. I'd say that's an error :)

Reply:
I'm not sure about this one. The error is returned on an exported method and mentions some package-internal implementation details. I've pushed a change in 1292ded544cbc14ffb0c8a37a56aa1548186b13c which does this. If you're happy with it, i'm ok to keep it.